aptos-move/framework/aptos-framework/sources/object.move

```diff
@@ -0,0 +1,433 @@
+    public fun derive_object_id(source: &address, seed: vector<u8>): ObjectId {
+        let bytes = bcs::to_bytes(source);
+        vector::append(&mut bytes, seed);
+        vector::push_back(&mut bytes, OBJECT_ADDRESS_SCHEME);
+        ObjectId { inner: from_bcs::to_address(hash::sha3_256(bytes)) }
+    }
+
+
+    /// Returns the address of an ObjectId
+    public fun object_id_address(object_id: &ObjectId): address {
+        object_id.inner
+    }
+
+    /// Create a new named object and return the CreatorRef. Named objects can be queried globally
+    /// by knowing the user generated seed used to create them. Named objects cannot be deleted.
+    public fun create_named_object(creator: &signer, seed: vector<u8>): CreatorRef {
+        let creator_address = signer::address_of(creator);
+        let id = derive_object_id(&creator_address, seed);
+        create_object_internal(creator_address, id, false)
+    }
+
+    public fun create_object_from_account(creator: &signer): CreatorRef {
+        let guid = account::create_guid(creator);
+        create_object_from_guid(signer::address_of(creator), guid)
+    }
+
+    public fun create_object_from_object(creator: &signer): CreatorRef acquires Object {
+        let guid = create_guid(creator);
+        create_object_from_guid(signer::address_of(creator), guid)
+    }
+
+    fun create_object_from_guid(creator_address: address, guid: guid::GUID): CreatorRef {
+        let bytes = bcs::to_bytes(&guid);
+        vector::push_back(&mut bytes, OBJECT_ADDRESS_SCHEME);
+        let object_id = ObjectId { inner: from_bcs::to_address(hash::sha3_256(bytes)) };
+        create_object_internal(creator_address, object_id, true)
+    }
+
+    fun create_object_internal(
+        creator_address: address,
+        id: ObjectId,
+        can_delete: bool,
+    ): CreatorRef {
+        assert!(!exists<Object>(id.inner), error::already_exists(EOBJECT_EXISTS));
+
+        let object_signer = create_signer(id.inner);
+        let guid_creation_num = 0;
+        let transfer_events_guid = guid::create(id.inner, &mut guid_creation_num);
+
+        move_to(
+            &object_signer,
+            Object {
+                guid_creation_num,
+                owner: creator_address,
+                allow_ungated_transfer: true,
+                transfer_events: event::new_event_handle(transfer_events_guid),
+            },
+        );
+        CreatorRef { self: id, can_delete }
+    }
+
+    // Creation helpers
+
+    public fun disallow_ungated_transfer(ref: &CreatorRef) acquires Object {
+        let object = borrow_global_mut<Object>(ref.self.inner);
+        object.allow_ungated_transfer = false;
+    }
+
+    public fun generate_delete_ref(ref: &CreatorRef): DeleteRef {
+        assert!(ref.can_delete, error::permission_denied(ECANNOT_DELETE));
+        DeleteRef { self: ref.self }
+    }
+
+    /// Create a signer for the CreatorRef
+    public fun generate_signer(ref: &CreatorRef): signer {
```

Comment:
would this expose this account to any resource write?

Reply:
not sure I follow, but this is only available to the creator or extender and they have permission to ... do whatever they want.